lib/OptimizationMOI/src/OptimizationMOI.jl

```diff
@@ -17,53 +20,63 @@ function SciMLBase.allowsconstraints(opt::Union{MOI.AbstractOptimizer,
     true
 end
 
-struct MOIOptimizationProblem{T, F <: OptimizationFunction, uType, P,
-                              JT <: DenseOrSparse{T}, HT <: DenseOrSparse{T},
-                              CHT <: DenseOrSparse{T}} <:
+struct MOIOptimizationCache{T, F <: OptimizationFunction, uType, P, LB, UB, I,
```

Comment:
What would such an `OptimizationCache` hold? Right now it is kind of big. It holds the problem, which holds the function which hold the system. Convinient but a nightmare in terms of interface.

Reply:
I think all the fields here, except remove the `prob` field for `OptimizationProblem` from this (since as I commented below we can also make the `build_solution` work directly with `OptimizationCache` then as you had suggested)

Reply:
> I think this can be a more general, `OptimizationCache` right? (and moved to SciMLBase) and then, exactly like how you have done here, we'd define `init` for it for each backend.

Why not stick the `OptimizationCache` into `Optimization.jl`? 

@ChrisRackauckas pointed out that a per-solver cache maybe needed in #352. I'm not quite sure yet.

Since `reinit(cache; p)` or something like `set_p(cache, p)` is the only real reason for a cache, it has to be possible to change the parameters.  All nonlinear sovers use `prob.f.f` and `prob.f.cons` which simply take `p` as an argument - easy! But what if the actual box constraints bounds are defined via parameters? 
I would like to implement an interface for linear solvers as well. The only way I see to implement this, is directly via the objective and constraints expressions and using `Symbolics.jl`.

Reply:
For now the `OptimizationSolution` does not support access of parameters, however it would be nice to provide this. Than we need `paramsyms`.

For the MOI case I think you added the `expr` and `cons_expr` to the `OptimizationFunction`, at least if the parameters are not replaced it might be usefull for the linear solver interface.

Reply:
>Why not stick the OptimizationCache into Optimization.jl?

Because afaik we keep all such structs in SciMLBase, so it avoids any recursive dependencies for these. For example in https://github.com/SciML/SciMLBase.jl/pull/286/files if instead of `C` you want to use `OptimizationCache` then it would have to be there since SciMLBase can't take Optimization as a dependency.

>@ChrisRackauckas pointed out that a per-solver cache may be needed in https://github.com/SciML/Optimization.jl/issues/352. I'm not quite sure yet.

Yeah me too, from what I can see, OptimizationCache seems like a union of OptimizationFunction and OptimizationProblem, such that it can store the complete state of the problem, hence it would end up needing all of the possible fields and we just initialize everything to `nothing` and fill it up as relevant for the backend.

>Since reinit(cache; p) or something like set_p(cache, p) is the only real reason for a cache, it has to be possible to change the parameters. All nonlinear sovers use prob.f.f and prob.f.cons which simply take p as an argument - easy! But what if the actual box constraints bounds are defined via parameters?

Sorry, not sure I understood this. Why would that not be possible? 🤔 

Regarding the expressions, yeah that would go into it as well, and replace any constant values interpolated in there if needed (but this can be punted for later depending on how much work it ends up being)

Reply:
> Because afaik we keep all such structs in SciMLBase, so it avoids any recursive dependencies for these.

`LinearCache` would be a counterexample but i can see that it might be beneficial.

If `ub` or `lb` are defined in terms of parameters of an MTK model, only the system knows how to compute this.
```julia
@parameters c
@variables x [bounds = (c, Inf)]
```
 It is stored in the `bounds` metadata of each variable. But accessing the metadata is only possible easily via MTK. We can of course forbid this in MTK.